ehrql/dataset_parameters.py

```diff
@@ -0,0 +1,95 @@
+import logging
+import sys
+from argparse import ArgumentParser
+from dataclasses import dataclass
+from typing import Any
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Parameter:
+    name: str
+    type: callable = str
+    default: Any | None = None
+
+
+def get_dataset_parameter(name, type: callable = str, default: Any | None = None):  # NOQA: A002
+    """
+    Retrieve dataset definition parameters that have been passed on the command line.
+
+    Dataset parameters are passed after a `--`, in the format `--key value`
+
+    For example, to pass values for two parameters, `max_age` and `region`:
+
+    ```
+    generate-dataset /path/to/dataset_definiton/.py -- --max_age 70 --sex female
+    ```
+
+    Multiple values can be passed per parameter, separated by a space, and will be parsed
+    as a list, e.g.
+    ```
+    generate-dataset /path/to/dataset_definiton/.py -- --sex male female
+    ```
+
+    For each parameter, call `get_dataset_paramter()`, e.g.:
+    ```
+        sex = get_dataset_parameter(name="sex")
+    ```
+
+    Parameters are parsed as strings, with a default value `None`, unless a `type` and/or
+    default value is explicitly specified. `type` can be any callable that takes a single string argument;
+    e.g. to convert to an integer, and default to 100 if no value is passed:
+    ```
+        max_age = get_dataset_parameter(name="max_age", type=int, default=100)
+    ```
+
+    To use in a dataset definition:
+
+    ```
+    from ehrql import create_dataset, get_dataset_parameter
+
+    max_age = get_dataset_parameter(name="max_age", type=int, default=20)
+    sex = get_dataset_parameter(name="sex", default="male")
+
+    dataset = create_dataset()
+    age = patients.age_on("2025-01-01")
+
+    dataset.define_population((age < max_age) & (patients.sex==sex))
+    ```
+    """
+    # find matching user args
```

Comment:
By default argparse supports both `--name value` and `--name=value` syntax and I think it might be helpful if we could do the same. We might get this for free if we were able to use `parse_known_args()` (see note below).